File: 02/Day02.java

import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Scanner;

public class Day02 {
    private static ArrayList<String> parseData(String fileName) {
        try {
            ArrayList<String> data = new ArrayList<>();
            File fl = new File(fileName);
            Scanner rdr = new Scanner(fl);

            while (rdr.hasNextLine()) {
                data.add(rdr.nextLine());
            }

            rdr.close();

            return data;

        } catch (FileNotFoundException e) {
            System.out.println("File not found");
        }

        return null;
    }

    private static Integer puzzle1(ArrayList<String> data) {
        HashMap<String, Integer> shapePoints = new HashMap<>();
        shapePoints.put("X", 1);
        shapePoints.put("Y", 2);
        shapePoints.put("Z", 3);

        HashMap<String, Integer> roundPoints = new HashMap<>();
        roundPoints.put("Win", 6);
        roundPoints.put("Draw", 3);
        roundPoints.put("Lost", 0);

        HashMap<String, Integer> outcomes = new HashMap<>();
        
        outcomes.put("A X", shapePoints.get("X") + roundPoints.get("Draw"));
        outcomes.put("A Y", shapePoints.get("Y") + roundPoints.get("Win"));
        outcomes.put("A Z", shapePoints.get("Z") + roundPoints.get("Lost"));

        outcomes.put("B X", shapePoints.get("X") + roundPoints.get("Lost"));
        outcomes.put("B Y", shapePoints.get("Y") + roundPoints.get("Draw"));
        outcomes.put("B Z", shapePoints.get("Z") + roundPoints.get("Win"));

        outcomes.put("C X", shapePoints.get("X") + roundPoints.get("Win"));
        outcomes.put("C Y", shapePoints.get("Y") + roundPoints.get("Lost"));
        outcomes.put("C Z", shapePoints.get("Z") + roundPoints.get("Draw"));

        return data.stream().mapToInt(line -> outcomes.get(line) ).sum();

    }

    private static Integer puzzle2(ArrayList<String> data) {
        HashMap<String, Integer> shapePoints = new HashMap<>();
        shapePoints.put("Rock", 1);
        shapePoints.put("Paper", 2);
        shapePoints.put("Scissors", 3);

        HashMap<String, Integer> roundPoints = new HashMap<>();
        roundPoints.put("Win", 6);
        roundPoints.put("Draw", 3);
        roundPoints.put("Lost", 0);

        HashMap<String, Integer> outcomes = new HashMap<>();
    
        outcomes.put("A X", roundPoints.get("Lost") + shapePoints.get("Scissors"));
        outcomes.put("B X", roundPoints.get("Lost") + shapePoints.get("Rock"));
        outcomes.put("C X", roundPoints.get("Lost") + shapePoints.get("Paper"));
    
        outcomes.put("A Y", roundPoints.get("Draw") + shapePoints.get("Rock"));
        outcomes.put("B Y", roundPoints.get("Draw") + shapePoints.get("Paper"));
        outcomes.put("C Y", roundPoints.get("Draw") + shapePoints.get("Scissors"));
    
        outcomes.put("A Z", roundPoints.get("Win") + shapePoints.get("Paper"));
        outcomes.put("B Z", roundPoints.get("Win") + shapePoints.get("Scissors"));
        outcomes.put("C Z", roundPoints.get("Win") + shapePoints.get("Rock"));

        return data.stream().mapToInt(outcomes::get).sum();
    }
    
    public static void main(String[] args) {
        ArrayList<String> sampleData = parseData("./day02sample.txt");
        ArrayList<String> data = parseData("./day02.txt");

        if(sampleData == null || data == null) {
            return;
        }

        System.out.println("puzzle1");
        System.out.println(puzzle1(sampleData));
        System.out.println(puzzle1(data));

        System.out.println("puzzle2");
        System.out.println(puzzle2(sampleData));
        System.out.println(puzzle2(data));
    }
}

File: 02/day02.ts
import * as fs from "fs";

const getData = (fileName:string):Array<Array<string>> => {
    const data:string = fs.readFileSync(fileName, 'utf-8').replace(/\r/g, '');

    const rounds = data.split('\n').map(round => round.split(' '));

    return rounds;
}


// A -> Rock, B -> Paper, C -> Scissors
// X -> Rock, Y -> Paper, Z -> Scissors

// Score = {Rock: 1, Paper: 2, Scissors: 3} + {Lost: 0, Draw: 3, Win: 6}

const shapePoints = {
    "Rock": 1,
    "Paper": 2,
    "Scissors": 3
};

const roundPoints = {
    "Lost": 0,
    "Draw": 3,
    "Win": 6
};

const otherMove = {
    "A": "Rock",
    "B": "Paper",
    "C": "Scissors"
};

const sampleData:Array<Array<string>> = getData('./day02sample.txt');
const data:Array<Array<string>> = getData('./day02.txt');

const puzzle1 = (data: Array<Array<string>>):number => {
    const calculateRoundScore = (round: Array<string>): number => {
        const selfMove = {
            "X": "Rock",
            "Y": "Paper",
            "Z": "Scissors"
        };
        
        const other = otherMove[round[0]];
        const self = selfMove[round[1]];
    
        let points = shapePoints[self];
        if(other === self) {
            // Tie
            points += roundPoints.Draw;
        } else {
            if((self === "Rock" && other === "Scissors") || (self === "Paper" && other === "Rock") || (self === "Scissors" && other === "Paper") ) {
                points += roundPoints.Win;
            } else {
                points += roundPoints.Lost;
            }
        }
    
        return points;
    }
    return data.reduce((total, round) => total + calculateRoundScore(round), 0);
}

const puzzle2 = (data: Array<Array<string>>):number => {
    const calculateRoundScore = (round: Array<string>): number => {
        const outcomes = {
            "X": "Lost",
            "Y": "Draw",
            "Z": "Win"
        };

        const other = otherMove[round[0]];
        const outcome = outcomes[round[1]];
        let self = "";

        let points = roundPoints[outcome];

        switch(outcome) {
            case "Draw":
                self = other;
                break;
            case "Win":
                switch(other) {
                    case "Scissors":
                        self = "Rock";
                        break;
                    case "Rock":
                        self = "Paper";
                        break;
                    case "Paper":
                        self = "Scissors";
                        break;
                }
                break;
            case "Lost":
                switch(other) {
                    case "Scissors":
                        self = "Paper";
                        break;
                    case "Rock":
                        self = "Scissors";
                        break;
                    case "Paper":
                        self = "Rock";
                        break;
                }
                break;
        }
        
        points += shapePoints[self];

        return points;
    }

    return data.reduce((total, round) => total + calculateRoundScore(round), 0);
}

console.log(puzzle1(sampleData));
console.time('puzzle1');
console.log(puzzle1(data));
console.timeEnd('puzzle1');

console.log(puzzle2(sampleData));
console.time('puzzle2');
console.log(puzzle2(data));
console.timeEnd('puzzle2');